### PR TITLE
Update Constant parsing to support more attributes

### PR DESCRIPTION
**Pull request by music-dino**
- Add parsing support for `value_float`, `value_floats`, `value_int`, `value_ints` attributes
- Disable failing tests
- Resolves migraphx-benchmark/AMDMIGraphX#76

For ONNX version 1.13 and above, certain tests failed due to lack of support for the aforementioned attributes.
After adding support, the same tests fail on account of CastLike being unsupported, so those tests are disabled until it is implemented.

### CODE REVIEW

File: src/onnx/parse_constant.cpp [hunk 1]
```diff
@@ -39,16 +39,29 @@ struct parse_constant : op_parser<parse_constant>
                           onnx_parser::node_info info,
                           const std::vector<instruction_ref>& /*args*/) const
     {
-        literal v = parser.parse_value(info.attributes.at("value"));
+        static const std::string attributes[] = {
+            "value", "value_float", "value_floats", "value_int", "value_ints"};
```
Comment (CharlieL7) [inline src/onnx/parse_constant.cpp:43]: style: use `std::vector<std::string>` instead

File: src/onnx/parse_constant.cpp [hunk 2]
```diff
@@ -39,16 +39,29 @@ struct parse_constant : op_parser<parse_constant>
...
+
+        auto&& attr = [&] {
+            for(const auto& a : attributes)
+            {
+                if(auto it = info.attributes.find(a); it != info.attributes.end())
+                {
+                    return it->second;
+                }
```
Comment (CharlieL7) [inline src/onnx/parse_constant.cpp:51]: style: use `contains(C, a)` instead from `ranges.hpp`

File: src/onnx/parse_constant.cpp [hunk 3]
```diff
@@ -39,16 +39,29 @@ struct parse_constant : op_parser<parse_constant>
...
+            }
+            MIGRAPHX_THROW("Constant does not contain any supported attribute");
+        }();
```
Comment (CharlieL7) [inline src/onnx/parse_constant.cpp:54]: Based on the ONNX spec it looks like having more than one value attribute set is also invalid. Probably want to also check that here.

File: src/onnx/parse_constant.cpp [hunk 4]
```diff
@@ -39,16 +39,29 @@ struct parse_constant : op_parser<parse_constant>
...
         // return empty literal
         if(v.get_shape().elements() == 0)
         {
             return info.add_literal(literal{v.get_shape().type()});
         }
 
-        auto dim_size = info.attributes.at("value").t().dims_size();
         // if dim_size is 0, it is a scalar
-        if(dim_size == 0)
+        if(attr.has_t() && attr.t().dims_size() == 0)
```
Comment (CharlieL7) [inline src/onnx/parse_constant.cpp:64]:
```suggestion
        if(attr.has_t() and attr.t().dims_size() == 0)
```
style change

File: src/onnx/parse_constant.cpp (hunk 3 above)
Comment (CharlieL7) [inline src/onnx/parse_constant.cpp:54]: style: why are we creating a lambda with capture by reference and then immediately calling it? The lambda doesn't look like it's used again. It looks like you could declare `attr` before and then set it.

Comment (CharlieL7) [inline src/onnx/parse_constant.cpp:54]: Should have `onnx_test.cpp` entries that trigger throws for code coverage

**Review by CharlieL7 — CHANGES_REQUESTED** [review]

File: src/onnx/parse_constant.cpp (hunk 1 above)
Comment (music-dino) [reply src/onnx/parse_constant.cpp]: Fixed

File: src/onnx/parse_constant.cpp (hunk 4 above)
Comment (music-dino) [reply src/onnx/parse_constant.cpp]: Fixed

File: src/onnx/parse_constant.cpp (hunk 2 above)
Comment (music-dino) [reply src/onnx/parse_constant.cpp]: Noted, the code uses it now.

File: src/onnx/parse_constant.cpp (hunk 3 above)
Comment (music-dino) [reply src/onnx/parse_constant.cpp]: Added checks for that case.

Comment (music-dino) [reply src/onnx/parse_constant.cpp]: This lambda has now been removed due to a rewrite.

Comment (music-dino) [reply src/onnx/parse_constant.cpp]: Tests cases for situations when the code is expected to throw have been added.

File: src/onnx/parse_constant.cpp [hunk 5]
```diff
@@ -39,16 +39,41 @@ struct parse_constant : op_parser<parse_constant>
...
+            std::string err = "Constant contains multiple attributes: " + *it;
+            for(auto i = std::next(it); i != attributes.end(); ++i)
+            {
+                if(p(*i))
+                {
+                    err += ", " + *i;
+                }
+            }
+            MIGRAPHX_THROW(err);
+        }
```
Comment (pfultz2) [inline src/onnx/parse_constant.cpp:64]: Use `copy_if` to copy all the attribute names to a vector and then if the vector > 1 throw an error about multiple attributes. You can use `join_strings` to convert the vector of strings to a comma-delimited string.

**Comment by CharlieL7** [comment]
It doesn't look like the code changes were pushed up yet?

**Comment by music-dino** [comment]
> It doesn't look like the code changes were pushed up yet?

I fail to see what you mean. The requested changes are present in the diff.

**Comment by CharlieL7** [comment]
> I fail to see what you mean. The requested changes are present in the diff.

Weird, I was unable to see the new changes earlier today. It looks like it was updated now

**Review by CharlieL7 — APPROVED** [review]
cppcheck error looks like a false positive from L60 of parse_constant.cpp.

File: src/onnx/parse_constant.cpp [hunk 6]
```diff
@@ -39,16 +40,37 @@ struct parse_constant : op_parser<parse_constant>
...
+            MIGRAPHX_THROW("Constant node does not contain any supported attribute");
+        }
+
+        if(present_attributes.size() > 1)
+        {
+            MIGRAPHX_THROW("Constant contains multiple attributes: " +
+                           join_strings(std::move(present_attributes), ", "));
+        }
+
+        auto&& attr = info.attributes[present_attributes[0]];
```
Comment (umangyadav) [inline src/onnx/parse_constant.cpp:63]: you can use `cppcheck-suppress` to avoid false positives.

**Review by umangyadav — APPROVED** [review]
Looks good to me

**Review by pfultz2 — APPROVED** [review]

**Comment by causten** [comment]
@music-dino  so close, can you fix the cppcheck

**Comment by music-dino** [comment]
> @music-dino so close, can you fix the cppcheck

Should be fixed now.

**Pull request closed without merging**